### DISCUSSION TITLE
How do I create a record owned by a user using client side JS?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#2481] [Q&A] [answered]
I have a logged in user. I want to create a record inside JavaScript (running inside the browser) that is owned by that user. I created a table that has a user relation. When I create the record, I believe I could do this by specifying the user ID when creating the record, right?

```
// example create data
const data = {
    "my_field": "test",
    "owner": "RELATION_RECORD_ID"
};

const record = await pb.collection('mytable').create(data);
```

But, this permits the user to assert who they are and I would rather not trust the client side to fill in this data. Is there a way to specify API rules such that this happens automatically, or at least, that a malicious user cannot specify another user's ID instead of their own?

This seems like the right API rule for create (right?), but I don't think gives me insight into how I could automatically populate the user:

```
@request.auth.id = @request.data.owner
```

Also, I'm actually "logging in" by modifying the handle to pocketbase manually: `pb.authStore.token = 'xxxx'`. This means I don't think I will have the ID of the user stored there, and might not be able to put it into the JS code anyway. I can work around this and manually set that as well, but it would be great if pocketbase could do that server side rather than client side.

**Comment by ganigeorgiev** [comment, marked as answer]
> I believe I could do this by specifying the user ID when creating the record, right?

Yes, you can pass directly the user id you want to assign.

The above API rule is correct. If `owner` can be empty you need to also ensure that `@request.auth.id` is not empty:

```js
@request.auth.id != "" && @request.auth.id = @request.data.owner
```

For update, if you want to allow only the owner of the record to be able to update its fields you can use something like:

```js
// allow access to only the owner
owner = @request.auth.id &&
// prevent new owner submission
@request.data.owner:isset = false
```
(_note that comments currently are not supported as part of the rules syntax_)

---

> how I could automatically populate the user

At the moment the only way to set the id of the current auth user server-side is to use the [`OnRecordBeforeCreateRequest` hook](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks), something like:

```go
app.OnRecordBeforeCreateRequest("users").Add(func(e *core.RecordCreateEvent) error {
	admin, _ := e.HttpContext.Get(apis.ContextAdminKey).(*models.Admin)
	if admin != nil {
		return nil // admins can submit anything
	}

	authRecord, _ := e.HttpContext.Get(apis.ContextAuthRecordKey).(*models.Record)
	if authRecord == nil {
		return apis.NewForbiddenError("Only authenticated users can perform this action", nil)
	}

	// always set the current authenticated record id as owner
	e.Record.Set("owner", authRecord.Id)

	return nil
})
```

**Reply by nqmgaming** [reply]
Can you update code to the latest version of pocketbase @xrd

**Reply by ganigeorgiev** [reply]
Here is a v0.23+ equivalent code sample:

```go
app.OnRecordCreateRequest("users").BindFunc(func(e *core.RecordRequestEvent) error {
    // superusers can submit anything
    if e.HasSuperuserAuth() {
        return e.Next()
    }

    if e.Auth == nil {
        return e.ForbiddenError("Only authenticated users can perform this action", nil)
    }

    // always set the current authenticated record id as owner
    e.Record.Set("owner", e.Auth.Id)

    return e.Next()
})
```

All `*Request` suffixed hooks embeds `core.RequestEvent`, meaning that all methods described in the routing docs should work - https://pocketbase.io/docs/go-routing/.

If you have other questions, please open a new Q&A discussion elaborating more on your use case and what actually you are trying to do.
